🪟🎉 Connector builder: Integrate connector form for test input

flash1293 opened the pull request:
## What

Closes https://github.com/airbytehq/airbyte/issues/17674

Integrate connector form
<img width="2087" alt="Screenshot 2022-12-12 at 18 06 33" src="https://user-images.githubusercontent.com/1508364/207139787-54488981-2dc8-426e-97bb-4ad858e5ceb5.png">


## How
I needed to extend the existing connector form in a bunch of ways to make this work:
* Introduce a new type for specification definitions which do not have an id (called them "SourceSpecificationDefinitionDraft"=
* Make the connector form work with SourceSpecificationDefinitionDraft and extend all downstream prop types to either work with SourceSpecificationDefinitionDraft or branch out before it hits code that relies on a specification id (mostly the auth code - I'm just not rendering auth buttons and there's a lot of work to be done throughout the system to make it work, so I think this limitation is OK for now)
* Pass in additional props to control the submit buttons: submit label, cancel callback and classname for the footer to style it correctly


On the connector builder side I render the form if there is a spec as part of the current manifest, otherwise it falls back to the existing JSON code editor.

## Recommended reading order
1. `airbyte-webapp/src/core/domain/connector/types.ts`
2. `airbyte-webapp/src/views/Connector/ConnectorForm/ConnectorForm.tsx`
3. Rest of changes in ` 
airbyte-webapp/src/views/Connector/ConnectorForm`
4. ` 
airbyte-webapp/src/views/Connector/ConnectorForm`

## 🚨 User Impact 🚨
Users can edit test inputs using the regular connector form instead of manually editing JSON

lmossman reviewed (commented):
Had a few comments but overall this is looking great.

I tested this out locally though and ran into a few issues that I think we'll want to address before merging:

I initially tried out adding this spec to the yaml editor:
```
spec:
  documentation_url: https://docsurl.com
  connection_specification:
    title: Poke API Spec
    type: object
    required:
      - pokemon_name
    additionalProperties: true
    properties:
      pokemon_name:
        type: string
        description: Pokemon name
```
and when I clicked the test input button, it caused the page to crash with the error:
```
schemaToFormBlock.ts:42 Uncaught TypeError: Cannot read properties of undefined (reading 'oneOf')
    at jsonSchemaToFormBlock (schemaToFormBlock.ts:42:1)
```
and I realized this was because that spec contained the property `connection_specification` instead of `connectionSpecification`, so when I fixed that it made the issue go away. But we should try make the code here more resilient to typos like that.

A few more things I noticed when testing:
- If I set the type of a field in the spec to `type: number`, then the User Inputs menu still renders a string-based input. This means there is no up/down arrow for adjusting the value, and it also incorrectly lets me enter a string value instead of restricting me to numeric characters. And when I do enter a string value, it prints out this error:
    ```
    react_devtools_backend.js:4012 getValues in useAuthentication failed. TypeError: The value of connectionConfiguration.pokemon_name could not be cast to a value that satisfies the schema type: "number". 
    ```
    This should be fixed so it actually renders a numeric input based on the type in the spec.
- If I first set the type of a field to string, open the menu and enter a string value, then close it and change the value to number, it still retains the string value and correctly doesn't let me submit. But I think the behavior here should probably be that it clears out the value when the type has changed (and it may be related to the above issue)
- I tried throwing the source-postgres spec into the yaml editor, and when I opened the test input menu it just had `...` for each conditional dropdown, rather than defaulting to the first option
- Another issue with the source-postgres spec was that it got too big to fit vertically on the screen, so it just got cut off on the top and bottom (see screenshot below). So we should add some scrolling behavior to the modal to fix this
<img width="1000" alt="Screen Shot 2022-12-16 at 2 52 52 PM" src="https://user-images.githubusercontent.com/22731524/208201841-cdf3653d-ca9a-4b3f-b62a-b6e550c97c39.png">

flash1293 commented on airbyte-webapp/src/views/Connector/ConnectorForm/useAuthentication.tsx:
```diff
@@ -178,7 +182,7 @@ export const useAuthentication = (): AuthenticationHook => {
     () => [
       // Fields from `advancedAuth` connectors
-      ...(advancedAuth
+      ...(advancedAuth && !isSourceDefinitionSpecificationDraft(connectorSpec)
```
I didn't find a straightforward way to do so. This is an annoying property of hooks - they always need to be called in the same order, so you can't conditionally call them as you suggested without breaking the [rules of hooks](https://reactjs.org/docs/hooks-rules.html) which seems worse to me than putting this condition deeper into the tree.

I agree that it's annoying that it needs to be checked in a bunch of places, but Typescript is flagging all of them so we can't run into cases where code would compile fine, then crash when a draft specification is handed to it.

To avoid this, we could pull the check above the hook in a parent component, then not render the component if the definition is a draft. However, that would require us to refactor a bunch of components and we would need to pass the spec down as a prop because `useConnectorForm` can always return a draft.

Overall, while this solution isn't ideal it seemed like the best approach without bigger changes in the structure.

lmossman replied:
Got it, your explanation makes sense. It's a good callout that typescript will flag any place that we forgot to add a check for this, so that will at least ensure we are consistent.

I don't think doing a big refactor of many components is worth it just to remove some of these conditionals, so I'm in agreement with you here 👍

flash1293 commented:
> and when I clicked the test input button, it caused the page to crash with the error:
> But we should try make the code here more resilient to typos like that.

Valid point, there is a bunch of ways to run into this (e.g. the good old malformed oneOfs). I added an error boundary to catch all errors and show them in a generic way:
<img width="981" alt="Screenshot 2022-12-18 at 15 57 39" src="https://user-images.githubusercontent.com/1508364/208305360-f0f32e69-a681-45d0-9fd0-9634a7d7d37c.png">

> If I set the type of a field in the spec to type: number, then the User Inputs menu still renders a string-based input

Good catch, as this is a general issue not related to connector builder, let me split that out into a separate PR. I will create a new issue for it.

> If I first set the type of a field to string, open the menu and enter a string value, then close it and change the value to number, it still retains the string value and correctly doesn't let me submit

I noticed the same issue, but I don't like clearing out the key - especially when working with the yaml it could often erase still valid data (e.g. if the value is "123" and switching to integer it still works) and it will get a little hacky to track if a property is changed while the user is editing the yaml. What about the following:
* In https://github.com/airbytehq/airbyte/pull/20431 , I added a "reset" button to the input form which serves this use case to a certain degree - if you are changing the test input and editing the inputs at the same time, you can always end up in weird states.
* The submit button doesn't work because the `schema.cast` throws - we could catch this error and show it like a test connection error. I think this change would be good in general for the connector form because it makes it more robust in the presence of weird configuration states which we can never fully rule out. If you like that I can split it out into a separate PR as well.

> I tried throwing the source-postgres spec into the yaml editor, and when I opened the test input menu it just had ... for each conditional dropdown, rather than defaulting to the first option

Yeah, this is happening because it won't apply defaults. I have fixed this on the follow-up PR here: https://github.com/airbytehq/airbyte/pull/20431

> Another issue with the source-postgres spec was that it got too big to fit vertically on the screen, so it just got cut off on the top and bottom (see screenshot below). So we should add some scrolling behavior to the modal to fix this

Good point, done

flash1293 commented:
Also simplified a bit the state management - as it's not possible anymore to have an invalid json string we need to track, we can only keep a single state object with the actual object in it around.

lmossman reviewed (approved):
One small suggestion, otherwise this LGTM. I agree with your assessment of how to approach the modifications to the spec block when the test inputs already have values.

Tested locally and it looks to be working as expected
